File: SHM/P.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <semaphore.h>
#include <fcntl.h>
#include <unistd.h>

#define SHM_SIZE 1024

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Uso: %s <string> <numero>\n", argv[0]);
        return 1;
    }

    int num = atoi(argv[2]);
    int shm_id;
    char *shm_ptr;
    key_t key = ftok("shmfile", 65);

    shm_id = shmget(key, SHM_SIZE, 0666 | IPC_CREAT);
    shm_ptr = (char *)shmat(shm_id, NULL, 0);

    sem_t *sem = sem_open("/mysem", O_CREAT, 0644, 1);
    
    sem_wait(sem);

    if (shm_ptr[2] == '3' && num == 3) {
        snprintf(shm_ptr + 3, SHM_SIZE - 2, "%s", argv[1]);
        printf("Memoria modificada: %s\n", shm_ptr);
    } else {
        printf("El número no coincide con 3.\n");
    }

    sem_post(sem);

    shmdt(shm_ptr);

    sem_close(sem);
    sem_unlink("/mysem");

    return 0;
}

File: SHM.py/P.py
import mmap
import os
import sys
import posix_ipc

if len(sys.argv) != 3:
    print(f"Uso: {sys.argv[0]} <string> <numero>")
    sys.exit(1)

SHM_NAME = "/shmfile"
input_string = sys.argv[1]
num = int(sys.argv[2])

shm = posix_ipc.SharedMemory(SHM_NAME)
memory = mmap.mmap(shm.fd, shm.size)
os.close(shm.fd)

sem = posix_ipc.Semaphore("/mysem", posix_ipc.O_CREAT, initial_value=1)

sem.acquire()

memory.seek(2)
if memory.read(1) == b'3' and num == 3:
    memory.seek(3)
    memory.write(input_string.encode())
    memory.seek(0)
    print("Memoria modificada:", memory.read().decode())
else:
    print("El número no coincide con 3.")

sem.release()

memory.close()
